# Autolib Electric Car Sharing Service

## Importing The Required Libraries

In [262]:
# Pandas library
import pandas as pd

# Numpy library
import numpy as np

## Loading The Dataset

In [263]:
# url
url = 'http://bit.ly/autolib_dataset'
# Dataset
ElectricCar = pd.read_csv(url, encoding= 'latin1')

## Dataset Preview

In [264]:
ElectricCar.head(10)

Address  Cars  Bluecar counter  Utilib counter  \
0                 2 Avenue de Suffren     0                0               0   
1           145 Rue Raymond Losserand     6                6               0   
2    2 Avenue John Fitzgerald Kennedy     3                3               0   
3          51 Rue Eugène Oudiné     3                3               1   
4  6 avenue de la Porte de Champerret     3                3               0   
5                8 Boulevard Voltaire     0                0               0   
6                      37 rue Leblanc     0                0               0   
7       17 Rue des Luaps Prolongée     3                3               1   
8               34 avenue Jean Moulin     1                1               0   
9        41 boulevard de Rochechouart     6                6               0   

   Utilib 1.4 counter  Charge Slots Charging Status        City  \
0                   0             0     nonexistent       Paris   
1                   0             0     operational       Paris   
2                   2             0     operational  Le Bourget   
3                   0             1     operational       Paris   
4                   0             0     nonexistent       Paris   
5                   0             0     nonexistent       Paris   
6                   0             0     nonexistent       Paris   
7                   0             0     nonexistent    Nanterre   
8                   0             0     nonexistent       Paris   
9                   0             0     nonexistent       Paris   

                                   Displayed comment  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   
5                                                NaN   
6  Station en parking (niv -1), accès 37 rue L...   
7                                                NaN   
8                                                NaN   
9                                                NaN   

                                  ID  ... Scheduled at Slots  Station type  \
0                    paris-suffren-2  ...          NaN     2       station   
1         paris-raymondlosserand-145  ...          NaN     0       station   
2  lebourget-johnfitzgeraldkennedy-2  ...          NaN     1       station   
3              paris-eugeneoudine-51  ...          NaN     2       station   
4          paris-portedechamperret-6  ...          NaN     3       station   
5                   paris-voltaire-8  ...          NaN     4       station   
6      paris-citroencevennes-parking  ...          NaN     0       station   
7                  nanterre-luaps-17  ...          NaN     0       station   
8                paris-jeanmoulin-34  ...          NaN     4       station   
9               paris-anvers-parking  ...          NaN     0       station   

   Status Subscription status  year  month day hour minute  
0      ok         nonexistent  2018      4   8   11     43  
1      ok         nonexistent  2018      4   6    7     24  
2      ok         nonexistent  2018      4   3   20     14  
3      ok         nonexistent  2018      4   4    4     37  
4      ok         nonexistent  2018      4   8   17     23  
5      ok         nonexistent  2018      4   6    7      2  
6  closed         nonexistent  2018      4   8   18     20  
7      ok         nonexistent  2018      4   4   22     13  
8      ok         nonexistent  2018      4   2   22     58  
9      ok         nonexistent  2018      4   4   15      2  

[10 rows x 25 columns]

In [265]:
# Viewing all columns
ElectricCar.head(5).transpose()

0                            1  \
Address              2 Avenue de Suffren    145 Rue Raymond Losserand   
Cars                                   0                            6   
Bluecar counter                        0                            6   
Utilib counter                         0                            0   
Utilib 1.4 counter                     0                            0   
Charge Slots                           0                            0   
Charging Status              nonexistent                  operational   
City                               Paris                        Paris   
Displayed comment                    NaN                          NaN   
ID                       paris-suffren-2   paris-raymondlosserand-145   
Kind                             STATION                      STATION   
Geo point                 48.857, 2.2917           48.83126, 2.313088   
Postal code                        75015                        75014   
Public name              Paris/Suffren/2  Paris/Raymond Losserand/145   
Rental status                operational                  operational   
Scheduled at                         NaN                          NaN   
Slots                                  2                            0   
Station type                     station                      station   
Status                                ok                           ok   
Subscription status          nonexistent                  nonexistent   
year                                2018                         2018   
month                                  4                            4   
day                                    8                            6   
hour                                  11                            7   
minute                                43                           24   

                                                        2  \
Address                  2 Avenue John Fitzgerald Kennedy   
Cars                                                    3   
Bluecar counter                                         3   
Utilib counter                                          0   
Utilib 1.4 counter                                      2   
Charge Slots                                            0   
Charging Status                               operational   
City                                           Le Bourget   
Displayed comment                                     NaN   
ID                      lebourget-johnfitzgeraldkennedy-2   
Kind                                              STATION   
Geo point                            48.938103, 2.4286035   
Postal code                                         93350   
Public name          Le Bourget/John Fitzgerald Kennedy/2   
Rental status                                 operational   
Scheduled at                                          NaN   
Slots                                                   1   
Station type                                      station   
Status                                                 ok   
Subscription status                           nonexistent   
year                                                 2018   
month                                                   4   
day                                                     3   
hour                                                   20   
minute                                                 14   

                                                3  \
Address                51 Rue Eugène Oudiné   
Cars                                            3   
Bluecar counter                                 3   
Utilib counter                                  1   
Utilib 1.4 counter                              0   
Charge Slots                                    1   
Charging Status                       operational   
City                                        Paris   
Displayed comment                             NaN   
ID                     

## Data Understanding

### Dataframe Information

In [266]:
ElectricCar.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 25 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Address              5000 non-null   object
 1   Cars                 5000 non-null   int64 
 2   Bluecar counter      5000 non-null   int64 
 3   Utilib counter       5000 non-null   int64 
 4   Utilib 1.4 counter   5000 non-null   int64 
 5   Charge Slots         5000 non-null   int64 
 6   Charging Status      5000 non-null   object
 7   City                 5000 non-null   object
 8   Displayed comment    111 non-null    object
 9   ID                   5000 non-null   object
 10  Kind                 5000 non-null   object
 11  Geo point            5000 non-null   object
 12  Postal code          5000 non-null   int64 
 13  Public name          5000 non-null   object
 14  Rental status        5000 non-null   object
 15  Scheduled at         47 non-null     object
 16  Slots 

In [267]:
# Viewing hidden column information
ElectricCar.iloc[:, range(19,22)].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 3 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Subscription status  5000 non-null   object
 1   year                 5000 non-null   int64 
 2   month                5000 non-null   int64 
dtypes: int64(2), object(1)
memory usage: 117.3+ KB


## Data Preperation

### Validity

#### 1. Removal of unimportant columns

In [268]:
# The Cars column is a duplicate of the Bluecar counter, therefore it is not required.
# The Geo point column is also unnecessary as the Address, Public name and the ID column cater to location requirements.
# The Address and Public name information column will aslo be dropped as they contain similar information to the ID column.
# They also have foreign characters which will be cumbersome to correct.

# List to be used for dropping the specified columns.
droplist = ['Cars', 'Geo point', 'Address', 'Public name']

In [269]:
# 2. Checking the validity of the 'Displayed comment' column.

ElectricCar['Displayed comment'].unique()

array([nan, 'Station en parking (niv -1), accÃ\x83¨s 37 rue Leblanc',
       'Station en parking (niv 0), accÃ\x83¨s 26 villa Croix Nivert',
       'Station en parking (niv -2), accÃ\x83¨s 4 av. Foch',
       "Borne d'abonnement en sous-sol (niv -1), accÃ\x83¨s 4 rue Lobau",
       'Station en parking (niv -1), accÃ\x83¨s rue Wilson',
       'Station en parking (niv -1), accÃ\x83¨s Place Georges Pompidou',
       'Station en parking (niv -3), accÃ\x83¨s 32 rue dâÂ\x80Â\x99Alsace',
       'Station en parking (niv -2), accÃ\x83¨s 2 avenue Gabriel',
       'Station en parking (niv -1), accÃ\x83¨s av. de Versailles',
       'Station en parking (niv -1), accÃ\x83¨s rue Abel Gance',
       'Station en parking (niv -1), accÃ\x83¨s 4 av. des Ternes',
       'Station en parking (niv -1), Station B (sur la droite)',
       'Station en parking',
       'Station en parking (niv -2), accÃ\x83¨s 125 bvd du Montparnasse',
       'Station en parking (niv -1), Station A (sur la gauche)',
 

In [270]:
# The column does not provide a significant contribution to the study.

droplist.append('Displayed comment',)

In [271]:
# 3. The study is limited to understanding car usage over time, therefore, the column can be removed.
# It also has limited information, compared to the amount of data available.

droplist.append('Scheduled at')

In [272]:
# Dropping all specified columns

ElectricCar.drop(droplist, axis = 1, inplace = True)

In [273]:
# Checking dataframe information after changes have been made.
ElectricCar.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 19 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Bluecar counter      5000 non-null   int64 
 1   Utilib counter       5000 non-null   int64 
 2   Utilib 1.4 counter   5000 non-null   int64 
 3   Charge Slots         5000 non-null   int64 
 4   Charging Status      5000 non-null   object
 5   City                 5000 non-null   object
 6   ID                   5000 non-null   object
 7   Kind                 5000 non-null   object
 8   Postal code          5000 non-null   int64 
 9   Rental status        5000 non-null   object
 10  Slots                5000 non-null   int64 
 11  Station type         5000 non-null   object
 12  Status               5000 non-null   object
 13  Subscription status  5000 non-null   object
 14  year                 5000 non-null   int64 
 15  month                5000 non-null   int64 
 16  day   

#### 2. Accuracy

##### a) Charge Slots and Charging Status

In [274]:
# Charge slots can only be greater than 0 if the Charging status is operational
# 1. Checking the condition is met

wrong_number_slots = ElectricCar[(ElectricCar['Charging Status'] == 'operational') & (ElectricCar['Charge Slots'] == 0)]
wrong_number_slots.any().any()


True

In [275]:
# Checking for value to replace the contradictory value.
ElectricCar[(ElectricCar['Charging Status'] == 'operational') & (ElectricCar['Charge Slots'] != 0)]['Charge Slots'].mode()

0    1
dtype: int64

In [276]:
# Implementing the required changes
ElectricCar.loc[(ElectricCar['Charging Status'] == 'operational') & (ElectricCar['Charge Slots'] == 0),'Charge Slots'] = 1
# Ensuring that changes have been made
wrong_number_slots.any().any()

True

In [277]:
# 2. Ensuring that the charging slots are only >1 when the status is operational.

wrong_slots = ElectricCar[(ElectricCar['Charging Status'] != 'operational') & (ElectricCar['Charge Slots'] != 0)]
wrong_slots.any().any()

False

In [278]:
# 3. Ensuring that the slots is zero for a charging status of future.

future_slots = ElectricCar[(ElectricCar['Charging Status'] == 'future') & (ElectricCar['Charge Slots'] != 0) & (ElectricCar['Slots'] != 0)]
future_slots.any().any()

False

In [279]:
ElectricCar.columns

Index(['Bluecar counter', 'Utilib counter', 'Utilib 1.4 counter',
       'Charge Slots', 'Charging Status', 'City', 'ID', 'Kind', 'Postal code',
       'Rental status', 'Slots', 'Station type', 'Status',
       'Subscription status', 'year', 'month', 'day', 'hour', 'minute'],
      dtype='object')

In [280]:
# 4. Ensuring that the cars are zero for a charging status of future.

future_cars = ElectricCar[(ElectricCar['Charging Status'] == 'future') & (ElectricCar['Bluecar counter'] != 0) &
(ElectricCar['Utilib counter'] != 0) & (ElectricCar['Utilib 1.4 counter'] != 0)]
future_cars.any().any()

False

The conditions have been met.

##### c) Kind Column

In [281]:
# 1. Ensuring that the 'CENTER' value has no resources

center = ElectricCar[(ElectricCar['Kind'] == 'CENTER') & (ElectricCar['Bluecar counter'] != 0) &
(ElectricCar['Utilib counter'] != 0) & (ElectricCar['Utilib 1.4 counter'] != 0) & ElectricCar['Charge Slots'] != 0 & (ElectricCar['Slots'] != 0)]
center.any().any()

False

In [282]:
# 2. Ensuring that the 'PARKING' value has no charging slots.

parking = ElectricCar[(ElectricCar['Kind'] == 'PARKING') & ElectricCar['Charge Slots'] != 0]
parking.any().any()

False

##### d) Rental Status

In [283]:
# Ensuring that the 'broken' status only has slot resources available.

broken= ElectricCar[(ElectricCar['Kind'] == 'broken')  & (ElectricCar['Bluecar counter'] != 0) &
(ElectricCar['Utilib counter'] != 0) & (ElectricCar['Utilib 1.4 counter'] != 0) & (ElectricCar['Charge Slots'] != 0)]
broken.any().any()

False

##### e) Status

In [284]:
# Ensuring that no resources are available for the "sceduled" status.

status = ElectricCar[(ElectricCar['Status'] == 'scheduled')  & (ElectricCar['Bluecar counter'] != 0) &
(ElectricCar['Utilib counter'] != 0) & (ElectricCar['Utilib 1.4 counter'] != 0) & (ElectricCar['Charge Slots'] != 0) & (ElectricCar['Slots'] != 0)]
status.any().any()

False

#### 3. Completeness

In [285]:
# Checking for missing values
ElectricCar.isnull().sum().sum()

0

There are no missing values.

#### 4. Consistency

In [286]:
# Checking for duplicates
ElectricCar.duplicated().sum()

0

There are no duplicates.

#### 5. Uniformity

##### a) Renaming Columns

In [287]:
# Changing the case of the columns to sentence case, as specified in the provided docummentaion.
# Function used to capitalize the column names
def SentenceCase(table):
    new_columns = []
    columns = list(table.columns.values)
    for x in columns:
        new_columns.append(x.title())
    table.columns = new_columns

# Function call
SentenceCase(ElectricCar)


In [288]:
# Check changes
ElectricCar.columns

Index(['Bluecar Counter', 'Utilib Counter', 'Utilib 1.4 Counter',
       'Charge Slots', 'Charging Status', 'City', 'Id', 'Kind', 'Postal Code',
       'Rental Status', 'Slots', 'Station Type', 'Status',
       'Subscription Status', 'Year', 'Month', 'Day', 'Hour', 'Minute'],
      dtype='object')

##### b) Data Type Conversion

In [289]:
# Checking assigned data types
ElectricCar.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 19 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Bluecar Counter      5000 non-null   int64 
 1   Utilib Counter       5000 non-null   int64 
 2   Utilib 1.4 Counter   5000 non-null   int64 
 3   Charge Slots         5000 non-null   int64 
 4   Charging Status      5000 non-null   object
 5   City                 5000 non-null   object
 6   Id                   5000 non-null   object
 7   Kind                 5000 non-null   object
 8   Postal Code          5000 non-null   int64 
 9   Rental Status        5000 non-null   object
 10  Slots                5000 non-null   int64 
 11  Station Type         5000 non-null   object
 12  Status               5000 non-null   object
 13  Subscription Status  5000 non-null   object
 14  Year                 5000 non-null   int64 
 15  Month                5000 non-null   int64 
 16  Day   

No columns require type conversion.

#### 6. Outliers

The numerical columns are restricted to a specific range of values, therefore, there was no need to remove any outliers.

## Analysis

**The analysis has been limited to the city of Paris, in the month of April 2018.**

### 1. Most popular hour of the day for picking up a shared electric car in the city of Paris over the month of April 2018.

In [290]:
# Function used to obtain the most popular picking up hour.
# The hours were ordered to obtain the change in the number of cars at a a station with time.

def PopularInApril(column):
    # Restricting analysis to Paris, in the month of April 2018.
    popular_hour = ElectricCar[(ElectricCar['City'] == 'Paris') & (ElectricCar['Month'] == 4) &
    # Grouping by the hour in descending order, to obtain the total number of Bluecars present at stations, per hour.
    (ElectricCar['Year'] == 2018)][['Hour', column]].groupby('Hour').sum().sort_values('Hour')
    # Obtaining the difference between the number of cars at stations for seccessive hours.
    popular_hour['Difference'] = popular_hour[column].diff()
    # The first difference will be null, therefore it will be filled with a zero.
    popular_hour.fillna(0, inplace= True)
    # The minimum difference corresponds to the highest number of car pick-ups.
    return popular_hour[popular_hour['Difference'] == popular_hour['Difference'].min()]

##### Bluecar

In [291]:
# Popular pick up hour for Bluecars.

bluecar_pickup = list(PopularInApril("Bluecar Counter").index)[0]
bluecar_pickup

22

The most popular picking hour for the Bluecar was 22:00 PM.

##### Utilib

In [292]:
# Popular pickup hour for Utilib cars.

utilib_pickup = list(PopularInApril("Utilib Counter").index)[0]
utilib_pickup

9

The most popular picking hour for the Utlilib car was 9:00 AM.

##### Utilib 1.4

In [293]:
# Popular pickup hour for Utilib 1.4 cars.

utilib14_pickup = list(PopularInApril("Utilib 1.4 Counter").index)[0]
utilib14_pickup

11

The most popular picking hour for the Utilib 1.4 car was 11:00 AM.

### 2. What is the most popular hour for returning cars?

In [294]:
# Function that takes in the required columns and returns the most popular hour for returning cars.

def PopularReturnHour(column):
    # Restricting analysis to Paris, in the month of April 2018.
    popular_hour = ElectricCar[(ElectricCar['City'] == 'Paris') & (ElectricCar['Month'] == 4) &
    # Grouping by the hour in descending order, to obtain the total number of Bluecars present at stations, per hour.
    (ElectricCar['Year'] == 2018)][['Hour', column]].groupby('Hour').sum().sort_values('Hour')
    # Obtaining the difference between the number of cars at stations for seccessive hours.
    popular_hour['Difference'] = popular_hour[column].diff()
    # The first difference will be null, therefore it will be filled with a zero.
    popular_hour.fillna(0, inplace= True)
    # The maximum difference corresponds to the highest number of car returns.
    return popular_hour[popular_hour['Difference'] == popular_hour['Difference'].max()]

##### Bluecar

In [295]:
# Popular hour for returning Bluecars.

bluecar_return = list(PopularReturnHour("Bluecar Counter").index)[0]
bluecar_return

21

The most popular return hour for the Bluecar was 21:00 PM.

##### Utilib

In [296]:
# Popular hour for returning Utilib cars.

utilib_return = list(PopularReturnHour("Utilib Counter").index)[0]
utilib_return

6

The most popular return hour for the Utilib car was 6:00 AM.

##### Utilib 1.4

In [297]:
# Popular hour for returning Utilib 1.4 cars.

utilib14_return = list(PopularReturnHour("Utilib 1.4 Counter").index)[0]
utilib14_return

21

The most popular return hour for the Utilib 1.4 car was 21:00 PM.

### 3. What station is the most popular?

#### a) Overall?


In [298]:
# Function that determines the most popular station.
# It is assumed that the station with the highest number of cars is the most popular.

def OverallPopularStation(column):
    # Restricting the analysis to the city of Paris, in the month of April 2018.
    popular_station_overall = ElectricCar[(ElectricCar['City'] == 'Paris') & (ElectricCar['Month'] == 4) &
    # Grouping by Id to get the total number of cars at various stations.
    (ElectricCar['Year'] == 2018)][['Id',column]].groupby('Id').sum()
    # Obtaining the station with the largest number of cars.
    return popular_station_overall[popular_station_overall[column] == popular_station_overall[column].max()]

##### Bluecar

In [299]:
# Most popular station for Bluecars.

popular_station_bluecar = list(OverallPopularStation("Bluecar Counter").index)
popular_station_bluecar

['paris-philippeauguste-126']

The most popular station for Bluecars, overall, had an Id of paris-philippeauguste-126.

##### Utilib

In [300]:
# Most popular station for Utilib cars.

popular_station_utilib = list(OverallPopularStation("Utilib Counter").index)
popular_station_utilib

['paris-macmahon-5', 'paris-portedemontrouge-8']

The most popular station for Utilib cars, overall, had Id's of paris-macmahon-5 and paris-portedemontrouge-8.

##### Utilib 1.4

In [301]:
# Most popular station for Utilib 1.4.

popular_station_utilib14 = list(OverallPopularStation("Utilib 1.4 Counter").index)
popular_station_utilib14

['paris-portedemontrouge-8']

The most popular station for Utilib 1.4 cars, overall, had an Id of paris-portedemontrouge-8.

#### b) At the most popular picking hour?

In [302]:
# Funtion that determines the most popular station at the most popular picking hour, for each respective car.
def PopularStationHour(column,hour):
    # Restricting the analysis to the city of Paris, in the month of April 2018.
    pop_station_pop_hour = ElectricCar[(ElectricCar['City'] == 'Paris') & (ElectricCar['Month'] == 4) &
    # Selecting the most popular hour for picking up cars and grouping by Id to get the total number of cars at various stations.
    (ElectricCar['Year'] == 2018) & (ElectricCar['Hour'] == hour)][['Id',column]].groupby('Id').sum()
    # Returning the most populat station at the most popular hour.
    return pop_station_pop_hour[pop_station_pop_hour[column] == pop_station_pop_hour[column].max()]

##### Bluecar

In [303]:
# Most popular station for Bluecars at the most popular pick up hour.

bluecar_popular_station_hour = list(PopularStationHour("Bluecar Counter", bluecar_pickup).index)
bluecar_popular_station_hour 

['paris-ravignan-1']

Most popular station for Bluecars at the most popular pick up hour had an Id of paris-ravignan-1.

##### Utilib

In [304]:
# Most popular station for Utilib cars at the most popular pick up hour.

utilib_popular_station_hour = list(PopularStationHour("Utilib Counter", utilib_pickup).index)
utilib_popular_station_hour 

['paris-fredericloliee-5',
 'paris-macmahon-5',
 'paris-marcadet-150',
 'paris-picpus-7',
 'paris-victoria-17']

Most popular station for Utilib cars at the most popular pick up hour had Ids of paris-fredericloliee-5, paris-macmahon-5, paris-marcadet-150, paris-picpus-7, and 
paris-victoria-17.

##### Utilib 1.4

In [305]:
# Most popular station for Utilib 1.4 cars at the most popular pick up hour.

utilib14_popular_station_hour = list(PopularStationHour("Utilib 1.4 Counter", utilib14_pickup).index)
utilib14_popular_station_hour 

['paris-berthier-97',
 'paris-federation-10',
 'paris-malesherbes-7',
 'paris-mathis-35',
 'paris-monge-58',
 'paris-parmentier-2',
 'paris-tolbiac-145',
 'paris-tolbiac-15',
 'paris-tolbiac-41',
 'paris-vaugirard-20']

Most popular station for Utilib 1.4 cars at the most popular pick up hour had Ids of paris-berthier-97, paris-federation-10, paris-malesherbes-7, paris-mathis-35,  paris-monge-58, paris-parmentier-2, paris-tolbiac-145, paris-tolbiac-15, paris-tolbiac-41, and paris-vaugirard-20.

### 4. What postal code is the most popular postal code for picking up cars? Does the most popular station belong to that postal code?

#### a) Overall?

In [306]:
# Function used to obtain the most popular code. 
# It is assumed that the most popula postal code has the highest number of cars.

def OverallPopularPostalCode(column):
    # Restricting the analysis to the city of Paris, in the month of April 2018.
    popular_postal = ElectricCar[(ElectricCar['City'] == 'Paris') & (ElectricCar['Month'] == 4) &
    # Grouping by Postal code to get the total number of cars at various postal codes.
    (ElectricCar['Year'] == 2018)][['Postal Code',column]].groupby('Postal Code').sum()
    # Returning the most popular postal codes for various car types.
    return popular_postal[popular_postal[column] == popular_postal[column].max()]
    

In [307]:
# Function to determine if the most popular station is in the most popular postal code.

def StationInPostalCode(station, code):
    present = []
    for codeelement in code:
        ids = list(ElectricCar[(ElectricCar['Postal Code'] == codeelement)]['Id'])
        for stationelement in station:
            if stationelement in ids:
                location = str(codeelement) + ' ' + stationelement
                present.append(location)
    return present



##### Bluecar

In [308]:
# Determining the most popular Bluecar postal code.

bluecar_popular_postal_code = list(OverallPopularPostalCode("Bluecar Counter").index)
bluecar_popular_postal_code

[75015]

In [309]:
# Checking if the most popular station(s) is in the most popular postal code.

StationInPostalCode(popular_station_bluecar, bluecar_popular_postal_code)

[]

The most popular Bluecar postal code was 75015, the most popular station was not in this postal code.

##### Utilib

In [310]:
# Determining the most popular Utilib car postal code.

utilib_popular_postal_code = list(OverallPopularPostalCode("Utilib Counter").index)
utilib_popular_postal_code

[75015, 75016, 75017]

In [311]:
# Checking if the most popular station(s) is in the most popular postal code.

StationInPostalCode(popular_station_utilib, utilib_popular_postal_code)


['75017 paris-macmahon-5']

The most popular Utilib car postal codes were 75015, 75016, 75017, only 75017 has one of the most popular utilib stations, which was paris-macmahon-5.

##### Utilib 1.4

In [312]:
# Determining the most popular Utilib 1.4 car postal code.

utilib14_popular_postal_code = list(OverallPopularPostalCode("Utilib 1.4 Counter").index)
utilib14_popular_postal_code

[75017]

In [313]:
# Checking if the most popular station(s) is in the most popular postal code.

StationInPostalCode(popular_station_utilib14, utilib14_popular_postal_code)

[]

The most popular Utilib 1.4 car postal code was  75017, it did not have the popular stations.

#### b) At the most popular picking hour?

In [314]:
# Function used to determine the most popular postal code at the most popular picking hour.
def PopularPostalCodeAtPickingHour(column, hour):
    # Restricting the analysis to the city of Paris, in the month of April 2018.
    popular_postal = ElectricCar[(ElectricCar['City'] == 'Paris') & (ElectricCar['Month'] == 4) &
    # Grouping by Postal code to get the total number of cars at various postal codes.
    (ElectricCar['Year'] == 2018) & (ElectricCar['Hour'] == hour)][['Postal Code',column]].groupby('Postal Code').sum()
    # Returning the most popular postal codes for various car types at the most popular picking hour.
    return popular_postal[popular_postal[column] == popular_postal[column].max()]

In [315]:
# popular_hour = ElectricCar[(ElectricCar['Hour'] == 22)][['Postal Code', 'Bluecar Counter']].groupby('Postal Code').sum()
# popular_hour['Difference'] = popular_hour['Bluecar Counter'].diff()
# popular_hour.fillna(0, inplace= True)
# popular_hour.sort_values('Difference').head(1)

##### Bluecar

In [316]:
# Most popular postal code for Bluecars at their most popular pick-up time.

bluecar_popular_post_code_at_picking_hour = list(PopularPostalCodeAtPickingHour("Bluecar Counter", bluecar_pickup).index)
bluecar_popular_post_code_at_picking_hour

[75016]

In [317]:
# Checking if the most popular station(s) is in this popular postal code.

StationInPostalCode(popular_station_bluecar, bluecar_popular_post_code_at_picking_hour)

[]

Most popular postal code for Bluecars at their most popular pick-up time was 75016, the most popular station was not found in this postal code.

##### Utilib

In [318]:
# Most popular postal code for Utilib cars at their most popular pick-up time.

utilib_popular_post_code_at_picking_hour = list(PopularPostalCodeAtPickingHour("Utilib Counter", utilib_pickup).index)
utilib_popular_post_code_at_picking_hour

[75001, 75012, 75017, 75018, 75020]

In [319]:
# Checking if the most popular station(s) is in this popular postal code.

StationInPostalCode(popular_station_utilib, utilib_popular_post_code_at_picking_hour)

['75017 paris-macmahon-5']

Most popular postal codes for Utilib cars at their most popular pick-up time were 75001, 75012, 75017, 75018, 75020, only 75017 had one of the most popular stations, which was paris-macmahon-5.

##### Utilib 1.4

In [320]:
# Most popular postal code for Utilib 1.4 cars at their most popular pick-up time.

utilib14_popular_post_code_at_picking_hour = list(PopularPostalCodeAtPickingHour("Utilib 1.4 Counter", utilib14_pickup).index)
utilib14_popular_post_code_at_picking_hour

[75013]

In [321]:
# Checking if the most popular station(s) is in this popular postal code.

StationInPostalCode(popular_station_utilib14, utilib14_popular_post_code_at_picking_hour)

[]

Most popular postal code for Utilib 1.4 cars at their most popular pick-up time was 75013, the most popular station was not found in this postal code.